// antares-mysql-dumper/lib/tools/fieldTypes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IS_MULTI_SPATIAL = exports.SPATIAL = exports.BIT = exports.BLOB = exports.HAS_TIMEZONE = exports.DATETIME = exports.TIME = exports.DATE = exports.BOOLEAN = exports.FLOAT = exports.NUMBER = exports.TEXT_SEARCH = exports.ARRAY = exports.LONG_TEXT = exports.TEXT = void 0;
exports.TEXT = [
    'CHAR',
    'VARCHAR',
    'CHARACTER',
    'CHARACTER VARYING'
];
exports.LONG_TEXT = [
    'TEXT',
    'MEDIUMTEXT',
    'LONGTEXT',
    'JSON',
    'VARBINARY'
];
exports.ARRAY = [
    'ARRAY',
    'ANYARRAY'
];
exports.TEXT_SEARCH = [
    'TSVECTOR',
    'TSQUERY'
];
exports.NUMBER = [
    'INT',
    'TINYINT',
    'SMALLINT',
    'MEDIUMINT',
    'BIGINT',
    'DECIMAL',
    'NUMERIC',
    'INTEGER',
    'SMALLSERIAL',
    'SERIAL',
    'BIGSERIAL',
    'OID',
    'XID'
];
exports.FLOAT = [
    'FLOAT',
    'DECIMAL',
    'DOUBLE',
    'REAL',
    'DOUBLE PRECISION',
    'MONEY'
];
exports.BOOLEAN = [
    'BOOL',
    'BOOLEAN'
];
exports.DATE = ['DATE'];
exports.TIME = [
    'TIME',
    'TIME WITH TIME ZONE'
];
exports.DATETIME = [
    'DATETIME',
    'TIMESTAMP',
    'TIMESTAMP WITHOUT TIME ZONE',
    'TIMESTAMP WITH TIME ZONE'
];
// Used to check datetime fields only
exports.HAS_TIMEZONE = [
    'TIMESTAMP WITH TIME ZONE',
    'TIME WITH TIME ZONE'
];
exports.BLOB = [
    'BLOB',
    'TINYBLOB',
    'MEDIUMBLOB',
    'LONGBLOB',
    'BYTEA'
];
exports.BIT = [
    'BIT',
    'BIT VARYING'
];
exports.SPATIAL = [
    'POINT',
    'LINESTRING',
    'POLYGON',
    'GEOMETRY',
    'MULTIPOINT',
    'MULTILINESTRING',
    'MULTIPOLYGON',
    'GEOMCOLLECTION',
    'GEOMETRYCOLLECTION'
];
// Used to check multi spatial fields only
exports.IS_MULTI_SPATIAL = [
    'MULTIPOINT',
    'MULTILINESTRING',
    'MULTIPOLYGON',
    'GEOMCOLLECTION',
    'GEOMETRYCOLLECTION'
];
